src/Rules/Operators/InvalidBinaryOperationRule.php

```diff
@@ -108,6 +111,28 @@ public function processNode(Node $node, Scope $scope): array
 			->assignVariable($rightName, $rightType, $rightType);
 
 		if (!$scope->getType($newNode) instanceof ErrorType) {
+			if (
+				$node instanceof Node\Expr\AssignOp\Div
+				|| $node instanceof Node\Expr\AssignOp\Mod
+				|| $node instanceof Node\Expr\BinaryOp\Div
+				|| $node instanceof Node\Expr\BinaryOp\Mod
+			) {
+				$zeroType = new UnionType([new ConstantIntegerType(0), new ConstantFloatType(0.0)]);
+				if ($zeroType->isSuperTypeOf($rightType)->maybe()) {
```

Comment:
This logic should be implemented as a `$callback` for the RuleLevelHelper above instead. It will make sure this is correctly reported on level 2 (when it's always zero) and on level 7 (when it might be a zero).

Reply:
the callback passed to `RuleLevelHelper` is only invoked for union types and intersection types.
in our case we pass in a expr which is resolved into a IntegerRangeType, which means the callback does not have any effect in this context.

Reply:
I get it. But still I'm missing at least asking for `$this->reportMaybes` so that maybe-zero is reported only on level 7+.

Reply:
Hey, it'd still be useful if the RuleLeveLHelper eliminated some situations. I can see that in the tests you're currently asserting an error that shouldn't be asserted because it's on a BenevolentUnionType.

Reply:
I tried testing the benevolent case, but for some reason the benevolence is lost somewhere?

https://phpstan.org/r/c0d12375-468d-46a2-8778-fcedc0d5e26f

Reply:
ok, hopefully I got it right now. my understanding about the expected results are
- we should not error on benevolent unions containing float (no matter the `$checkBenevolentUnionTypes` flag)
- we should error on unions with int|float
- we should error on benevolent unions with non-float but number-castable types respecting `$checkBenevolentUnionTypes`